src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,112 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double COEFFICIENT = 1.5;
+    private T[] values;
+    private int size;
+
+    public ArrayList() {
+        values = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    private void resize() {
+        int newCapacity = (int) (values.length * COEFFICIENT);
+        T[] newArray = (T[]) new Object[newCapacity];
+        System.arraycopy(values, 0, newArray, 0, size);
+        values = newArray;
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == values.length) {
+            resize();
+        }
+        values[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndex(index, value);
+        if (size == values.length) {
+            resize();
+        }
+        System.arraycopy(values, index, values, index + 1, size - index);
+        values[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (list == null) {
+            throw new NullPointerException("List is null");
+        }
+        if (size + list.size() > values.length) {
+            int newCapacity = (int) (Math.max(size + list.size(), values.length * COEFFICIENT));
+            T[] newArray = (T[]) new Object[newCapacity];
+            System.arraycopy(values, 0, newArray, 0, size);
+            values = newArray;
+        }
+        for (T element : list) {
+            add(element);
+        }
```

Comment:
Resize the array in a separate method.
According to one of the important principles called Single Responsibility principle, any software component must have only one responsibility, that is, to be in charge of doing just one concrete thing.